### DISCUSSION TITLE
Ep126: inout input clarification

### DISCUSSION THREAD

**Discussion by filblue** [pointfreeco/episode-code-samples#80] [Q&A] [answered]
Hello. In [Episode #126 Generalized Parsing: Part 3](https://www.pointfree.co/episodes/ep126-generalized-parsing-part-3), the [`Parser.run(input:output:)`](https://github.com/pointfreeco/episode-code-samples/blob/845d7e7de8f2fe81f8751cbdd91265193340d895/0126-generalized-parsing-pt3/stdin/Sources/stdin/main.swift#L69) function has `inout` attribute for the `input` parameter. I was not able to figure out the rationale for this. I checked, and it seems to work the same with immutable input iterator. @stephencelis could you please give a hint, am I missing something here? Thank you.

**Comment by mbrandonw** [comment, marked as answer]
Hi @filblue, you can find our rationale for the `inout` parameter in one of our early parsing episodes: https://www.pointfree.co/collections/parsing/what-is-parsing/ep57-what-is-a-parser-part-2

It basically boils down to two things: Swift can be more performant by using `inout` and the ergonomics of using `inout` are better than returning a value.

**Reply by filblue** [reply]
Hi @mbrandonw, thank you for taking your time to answer! I totally see the point of using `inout` for things that are mutated (consumed) by the parser. So, the rationale you mentioned above is relevant to why `var buffer` is mutable and `inout`. However, I don't see that `input: inout AnyIterator<Input>` is ever mutated.

**Reply by mbrandonw** [reply]
Oh! Sorry, I completely misunderstood your question. I see you are specifically talking about the `AnyIterator` stream stuff.

We have that marked as `inout` because the `.next()` method on `IteratorProtocol` is marked as mutating:

```swift
mutating func next() -> Self.Element?
```

However it seems that when it's used on an `AnyIterator` you don't need to have a mutable value 🤔. I'm not really sure why, and don't have time to look into it now, but it does seem like we could maybe clean something up here.

**Reply by filblue** [reply]
```
protocol MyProtocol {
    mutating func f() -> Int
}

struct A: MyProtocol {
    var i = 0
    mutating func f() -> Int { i += 1; return i }
}

struct B: MyProtocol {
    func f() -> Int { return 0 }
}

var a: MyProtocol = A()
print(a.f()) // Ok, no surprises

let b0 = B()
print(b0.f()) // Ok, a little surprisingly

let b1: MyProtocol = B()
print(b1.f()) // Error: Cannot use mutating member on immutable value
```

**Reply by mbrandonw** [reply]
Ha, I think that's it! Looks like `AnyIterator`'s implementation of `IteratorProtocol` uses a non-mutating `next()`:

```swift
extension AnyIterator : IteratorProtocol {

    /// Advances to the next element and returns it, or `nil` if no next element
    /// exists.
    ///
    /// Once `nil` has been returned, all subsequent calls return `nil`.
    @inlinable public func next() -> Element?
}
```

Mystery solved!
